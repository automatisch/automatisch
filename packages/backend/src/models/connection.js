import AES from 'crypto-js/aes.js';
import enc from 'crypto-js/enc-utf8.js';
import App from './app.js';
import AppConfig from './app-config.js';
import OAuthClient from './oauth-client.js';
import Base from './base.js';
import User from './user.js';
import Step from './step.js';
import appConfig from '../config/app.js';
import Telemetry from '../helpers/telemetry/index.js';
import globalVariable from '../helpers/global-variable.js';
import NotAuthorizedError from '../errors/not-authorized.js';

class Connection extends Base {
  static tableName = 'connections';

  static jsonSchema = {
    type: 'object',
    required: ['key'],

    properties: {
      id: { type: 'string', format: 'uuid' },
      key: { type: 'string', minLength: 1, maxLength: 255 },
      data: { type: 'string' },
      formattedData: { type: 'object' },
      userId: { type: 'string', format: 'uuid' },
      oauthClientId: { type: 'string', format: 'uuid' },
      verified: { type: 'boolean', default: false },
      draft: { type: 'boolean' },
      deletedAt: { type: 'string' },
      createdAt: { type: 'string' },
      updatedAt: { type: 'string' },
    },
  };

  static relationMappings = () => ({
    user: {
      relation: Base.BelongsToOneRelation,
      modelClass: User,
      join: {
        from: 'connections.user_id',
        to: 'users.id',
      },
    },
    steps: {
      relation: Base.HasManyRelation,
      modelClass: Step,
      join: {
        from: 'connections.id',
        to: 'steps.connection_id',
      },
    },
    triggerSteps: {
      relation: Base.HasManyRelation,
      modelClass: Step,
      join: {
        from: 'connections.id',
        to: 'steps.connection_id',
      },
      filter(builder) {
        builder.where('type', '=', 'trigger');
      },
    },
    appConfig: {
      relation: Base.BelongsToOneRelation,
      modelClass: AppConfig,
      join: {
        from: 'connections.key',
        to: 'app_configs.key',
      },
    },
    oauthClient: {
      relation: Base.BelongsToOneRelation,
      modelClass: OAuthClient,
      join: {
        from: 'connections.oauth_client_id',
        to: 'oauth_clients.id',
      },
    },
  });

  encryptData() {
    if (!this.eligibleForEncryption()) return;

    this.data = AES.encrypt(
      JSON.stringify(this.formattedData),
      appConfig.encryptionKey
    ).toString();

    delete this.formattedData;
  }

  decryptData() {
    if (!this.eligibleForDecryption()) return;

    this.formattedData = JSON.parse(
      AES.decrypt(this.data, appConfig.encryptionKey).toString(enc)
    );
  }

  eligibleForEncryption() {
    return this.formattedData ? true : false;
  }

  eligibleForDecryption() {
    return this.data ? true : false;
  }

  async getApp() {
    if (!this.key) return null;

    return await App.findOneByKey(this.key);
  }

  async getAppConfig() {
    return await AppConfig.query().findOne({ key: this.key });
  }

  async checkEligibilityForCreation() {
    const app = await this.getApp();

    const appConfig = await this.getAppConfig();

    if (appConfig) {
      if (appConfig.disabled) {
        throw new NotAuthorizedError(
          'The application has been disabled for new connections!'
        );
      }

      if (appConfig.useOnlyPredefinedAuthClients && this.formattedData) {
        throw new NotAuthorizedError(
          `New custom connections have been disabled for ${app.name}!`
        );
      }

      if (!this.formattedData) {
        const authClient = await appConfig
          .$relatedQuery('oauthClients')
          .findById(this.oauthClientId)
          .where({ active: true })
          .throwIfNotFound();

        this.formattedData = authClient.formattedAuthDefaults;
      }
    }

    return this;
  }

  async testAndUpdateConnection() {
    const app = await this.getApp();
    const $ = await globalVariable({ connection: this, app });

    let isStillVerified;

    try {
      isStillVerified = !!(await app.auth.isStillVerified($));
    } catch {
      isStillVerified = false;
    }

    return await this.$query().patchAndFetch({
      formattedData: this.formattedData,
      verified: isStillVerified,
    });
  }

  async verifyAndUpdateConnection(request) {
    const app = await this.getApp();
    const $ = await globalVariable({ 
      connection: this, 
      app,
      request 
    });
    await app.auth.verifyCredentials($);

    return await this.$query().patchAndFetch({
      verified: true,
      draft: false,
    });
  }

  async verifyWebhook(request) {
    if (!this.key) return true;

    const app = await this.getApp();

    const $ = await globalVariable({
      connection: this,
      request,
    });

    if (!app.auth?.verifyWebhook) return true;

    return app.auth.verifyWebhook($);
  }

  async generateAuthUrl() {
    const app = await this.getApp();
    const $ = await globalVariable({ connection: this, app });

    await app.auth.generateAuthUrl($);

    const url = this.formattedData.url;

    return { url };
  }

  async reset() {
    const formattedData = this?.formattedData?.screenName
      ? { screenName: this.formattedData.screenName }
      : {};

    const updatedConnection = await this.$query().patchAndFetch({
      formattedData,
    });

    return updatedConnection;
  }

  async updateFormattedData({ formattedData, oauthClientId }) {
    if (oauthClientId) {
      const oauthClient = await OAuthClient.query()
        .findById(oauthClientId)
        .throwIfNotFound();

      formattedData = oauthClient.formattedAuthDefaults;
    }

    return await this.$query().patchAndFetch({
      formattedData: {
        ...this.formattedData,
        ...formattedData,
      },
    });
  }

  // TODO: Make another abstraction like beforeSave instead of using
  // beforeInsert and beforeUpdate separately for the same operation.
  async $beforeInsert(queryContext) {
    await super.$beforeInsert(queryContext);

    await this.checkEligibilityForCreation();

    this.encryptData();
  }

  async $beforeUpdate(opt, queryContext) {
    await super.$beforeUpdate(opt, queryContext);
    this.encryptData();
  }

  async $afterFind() {
    this.decryptData();
  }

  async $afterInsert(queryContext) {
    await super.$afterInsert(queryContext);
    Telemetry.connectionCreated(this);
  }

  async $afterUpdate(opt, queryContext) {
    await super.$afterUpdate(opt, queryContext);
    Telemetry.connectionUpdated(this);
  }
}

export default Connection;
