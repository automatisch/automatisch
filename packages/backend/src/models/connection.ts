import { QueryContext, ModelOptions } from 'objection';
import type { RelationMappings } from 'objection';
import { AES, enc } from 'crypto-js';
import { IRequest } from '@automatisch/types';
import App from './app';
import Base from './base';
import User from './user';
import Step from './step';
import ExtendedQueryBuilder from './query-builder';
import appConfig from '../config/app';
import { IJSONObject } from '@automatisch/types';
import Telemetry from '../helpers/telemetry';
import globalVariable from '../helpers/global-variable';

class Connection extends Base {
  id!: string;
  key!: string;
  data: string;
  formattedData?: IJSONObject;
  userId!: string;
  verified: boolean;
  draft: boolean;
  count?: number;
  flowCount?: number;
  user?: User;
  steps?: Step[];
  triggerSteps?: Step[];

  static tableName = 'connections';

  static jsonSchema = {
    type: 'object',
    required: ['key'],

    properties: {
      id: { type: 'string', format: 'uuid' },
      key: { type: 'string', minLength: 1, maxLength: 255 },
      data: { type: 'string' },
      formattedData: { type: 'object' },
      userId: { type: 'string', format: 'uuid' },
      verified: { type: 'boolean', default: false },
      draft: { type: 'boolean' },
      deletedAt: { type: 'string' },
      createdAt: { type: 'string' },
      updatedAt: { type: 'string' },
    },
  };

  static relationMappings = (): RelationMappings => ({
    user: {
      relation: Base.BelongsToOneRelation,
      modelClass: User,
      join: {
        from: 'connections.user_id',
        to: 'users.id',
      },
    },
    steps: {
      relation: Base.HasManyRelation,
      modelClass: Step,
      join: {
        from: 'connections.id',
        to: 'steps.connection_id',
      },
    },
    triggerSteps: {
      relation: Base.HasManyRelation,
      modelClass: Step,
      join: {
        from: 'connections.id',
        to: 'steps.connection_id',
      },
      filter(builder: ExtendedQueryBuilder<Step>) {
        builder.where('type', '=', 'trigger');
      },
    },
  });

  encryptData(): void {
    if (!this.eligibleForEncryption()) return;

    this.data = AES.encrypt(
      JSON.stringify(this.formattedData),
      appConfig.encryptionKey
    ).toString();

    delete this.formattedData;
  }

  decryptData(): void {
    if (!this.eligibleForDecryption()) return;

    this.formattedData = JSON.parse(
      AES.decrypt(this.data, appConfig.encryptionKey).toString(enc.Utf8)
    );
  }

  eligibleForEncryption(): boolean {
    return this.formattedData ? true : false;
  }

  eligibleForDecryption(): boolean {
    return this.data ? true : false;
  }

  // TODO: Make another abstraction like beforeSave instead of using
  // beforeInsert and beforeUpdate separately for the same operation.
  async $beforeInsert(queryContext: QueryContext): Promise<void> {
    await super.$beforeInsert(queryContext);
    this.encryptData();
  }

  async $beforeUpdate(
    opt: ModelOptions,
    queryContext: QueryContext
  ): Promise<void> {
    await super.$beforeUpdate(opt, queryContext);
    this.encryptData();
  }

  async $afterFind(): Promise<void> {
    this.decryptData();
  }

  async $afterInsert(queryContext: QueryContext) {
    await super.$afterInsert(queryContext);
    Telemetry.connectionCreated(this);
  }

  async $afterUpdate(opt: ModelOptions, queryContext: QueryContext) {
    await super.$afterUpdate(opt, queryContext);
    Telemetry.connectionUpdated(this);
  }

  async getApp() {
    if (!this.key) return null;

    return await App.findOneByKey(this.key);
  }

  async verifyWebhook(request: IRequest) {
    if (!this.key) return true;

    const app = await this.getApp();

    const $ = await globalVariable({
      connection: this,
      request,
    });

    if (!app.auth?.verifyWebhook) return true;

    return app.auth.verifyWebhook($);
  }
}

export default Connection;
