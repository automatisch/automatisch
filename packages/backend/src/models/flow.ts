import { ValidationError } from 'objection';
import type {
  ModelOptions,
  QueryContext,
  StaticHookArguments,
} from 'objection';
import ExtendedQueryBuilder from './query-builder';
import Base from './base';
import Step from './step';
import User from './user';
import Execution from './execution';
import Telemetry from '../helpers/telemetry';

class Flow extends Base {
  id!: string;
  name!: string;
  userId!: string;
  active: boolean;
  status: 'paused' | 'published' | 'draft';
  steps: Step[];
  triggerStep: Step;
  publishedAt: string;
  remoteWebhookId: string;
  executions?: Execution[];
  lastExecution?: Execution;
  user?: User;

  static tableName = 'flows';

  static jsonSchema = {
    type: 'object',
    required: ['name'],

    properties: {
      id: { type: 'string', format: 'uuid' },
      name: { type: 'string', minLength: 1 },
      userId: { type: 'string', format: 'uuid' },
      remoteWebhookId: { type: 'string' },
      active: { type: 'boolean' },
      publishedAt: { type: 'string' },
      deletedAt: { type: 'string' },
      createdAt: { type: 'string' },
      updatedAt: { type: 'string' },
    },
  };

  static relationMappings = () => ({
    steps: {
      relation: Base.HasManyRelation,
      modelClass: Step,
      join: {
        from: 'flows.id',
        to: 'steps.flow_id',
      },
      filter(builder: ExtendedQueryBuilder<Step>) {
        builder.orderBy('position', 'asc');
      },
    },
    triggerStep: {
      relation: Base.HasOneRelation,
      modelClass: Step,
      join: {
        from: 'flows.id',
        to: 'steps.flow_id',
      },
      filter(builder: ExtendedQueryBuilder<Step>) {
        builder
          .where('type', 'trigger')
          .limit(1)
          .first();
      },
    },
    executions: {
      relation: Base.HasManyRelation,
      modelClass: Execution,
      join: {
        from: 'flows.id',
        to: 'executions.flow_id',
      },
    },
    lastExecution: {
      relation: Base.HasOneRelation,
      modelClass: Execution,
      join: {
        from: 'flows.id',
        to: 'executions.flow_id',
      },
      filter(builder: ExtendedQueryBuilder<Execution>) {
        builder.orderBy('created_at', 'desc').limit(1).first();
      },
    },
    user: {
      relation: Base.HasOneRelation,
      modelClass: User,
      join: {
        from: 'flows.user_id',
        to: 'users.id',
      },
    },
  });

  static async afterFind(args: StaticHookArguments<any>): Promise<any> {
    const { result } = args;

    const referenceFlow = result[0];

    if (referenceFlow) {
      const shouldBePaused = await referenceFlow.isPaused();

      for (const flow of result) {
        if (!flow.active) {
          flow.status = 'draft';
        } else if (flow.active && shouldBePaused) {
          flow.status = 'paused';
        } else {
          flow.status = 'published';
        }
      }
    }
  }

  async lastInternalId() {
    const lastExecution = await this.$relatedQuery('lastExecution');

    return lastExecution ? (lastExecution as Execution).internalId : null;
  }

  async lastInternalIds(itemCount = 50) {
    const lastExecutions = await this.$relatedQuery('executions')
      .select('internal_id')
      .orderBy('created_at', 'desc')
      .limit(itemCount);

    return lastExecutions.map((execution) => execution.internalId);
  }

  async $beforeUpdate(
    opt: ModelOptions,
    queryContext: QueryContext
  ): Promise<void> {
    await super.$beforeUpdate(opt, queryContext);

    if (!this.active) return;

    const oldFlow = opt.old as Flow;

    const incompleteStep = await oldFlow.$relatedQuery('steps').findOne({
      status: 'incomplete',
    });

    if (incompleteStep) {
      throw new ValidationError({
        message: 'All steps should be completed before updating flow status!',
        type: 'incompleteStepsError',
      });
    }

    const allSteps = await oldFlow.$relatedQuery('steps');

    if (allSteps.length < 2) {
      throw new ValidationError({
        message:
          'There should be at least one trigger and one action steps in the flow!',
        type: 'insufficientStepsError',
      });
    }

    return;
  }

  async $afterInsert(queryContext: QueryContext) {
    await super.$afterInsert(queryContext);
    Telemetry.flowCreated(this);
  }

  async $afterUpdate(opt: ModelOptions, queryContext: QueryContext) {
    await super.$afterUpdate(opt, queryContext);
    Telemetry.flowUpdated(this);
  }

  async getTriggerStep(): Promise<Step> {
    return await this.$relatedQuery('steps').findOne({
      type: 'trigger',
    });
  }

  async isPaused() {
    const user = await this.$relatedQuery('user');
    const allowedToRunFlows = await user.isAllowedToRunFlows();
    return allowedToRunFlows ? false : true;
  }
}

export default Flow;
