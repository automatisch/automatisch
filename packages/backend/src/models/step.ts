import { URL } from 'node:url';
import { QueryContext, ModelOptions } from 'objection';
import get from 'lodash.get';
import type { IJSONObject, IStep } from '@automatisch/types';
import Base from './base';
import App from './app';
import Flow from './flow';
import Connection from './connection';
import ExecutionStep from './execution-step';
import Telemetry from '../helpers/telemetry';
import appConfig from '../config/app';

class Step extends Base {
  id!: string;
  flowId!: string;
  key?: string;
  appKey?: string;
  type!: IStep['type'];
  connectionId?: string;
  status: 'incomplete' | 'completed';
  position!: number;
  parameters: IJSONObject;
  connection?: Connection;
  flow: Flow;
  executionSteps: ExecutionStep[];
  webhookPath?: string;

  static tableName = 'steps';

  static jsonSchema = {
    type: 'object',
    required: ['type'],

    properties: {
      id: { type: 'string', format: 'uuid' },
      flowId: { type: 'string', format: 'uuid' },
      key: { type: ['string', 'null'] },
      appKey: { type: ['string', 'null'], minLength: 1, maxLength: 255 },
      type: { type: 'string', enum: ['action', 'trigger'] },
      connectionId: { type: ['string', 'null'], format: 'uuid' },
      status: {
        type: 'string',
        enum: ['incomplete', 'completed'],
        default: 'incomplete',
      },
      position: { type: 'integer' },
      parameters: { type: 'object' },
      webhookPath: { type: ['string', 'null'] },
      deletedAt: { type: 'string' },
      createdAt: { type: 'string' },
      updatedAt: { type: 'string' },
    },
  };

  static get virtualAttributes() {
    return ['iconUrl', 'webhookUrl'];
  }

  static relationMappings = () => ({
    flow: {
      relation: Base.BelongsToOneRelation,
      modelClass: Flow,
      join: {
        from: 'steps.flow_id',
        to: 'flows.id',
      },
    },
    connection: {
      relation: Base.HasOneRelation,
      modelClass: Connection,
      join: {
        from: 'steps.connection_id',
        to: 'connections.id',
      },
    },
    executionSteps: {
      relation: Base.HasManyRelation,
      modelClass: ExecutionStep,
      join: {
        from: 'steps.id',
        to: 'execution_steps.step_id',
      },
    },
  });

  get webhookUrl() {
    return new URL(this.webhookPath, appConfig.webhookUrl).toString();
  }

  get iconUrl() {
    if (!this.appKey) return null;

    return `${appConfig.baseUrl}/apps/${this.appKey}/assets/favicon.svg`;
  }

  async computeWebhookPath() {
    if (this.type === 'action') return null;

    const triggerCommand = await this.getTriggerCommand();

    if (!triggerCommand) return null;

    const {
      useSingletonWebhook,
      singletonWebhookRefValueParameter,
      type,
    } = triggerCommand;

    const isWebhook = type === 'webhook';

    if (!isWebhook) return null;

    if (singletonWebhookRefValueParameter) {
      const parameterValue = get(this.parameters, singletonWebhookRefValueParameter);
      return `/webhooks/connections/${this.connectionId}/${parameterValue}`;
    }

    if (useSingletonWebhook) {
      return `/webhooks/connections/${this.connectionId}`;
    }

    return `/webhooks/flows/${this.flowId}`;
  }

  async getWebhookUrl() {
    if (this.type === 'action') return;

    const path = await this.computeWebhookPath();
    const webhookUrl = new URL(path, appConfig.webhookUrl).toString();

    return webhookUrl;
  }

  async $afterInsert(queryContext: QueryContext) {
    await super.$afterInsert(queryContext);
    Telemetry.stepCreated(this);
  }

  async $afterUpdate(opt: ModelOptions, queryContext: QueryContext) {
    await super.$afterUpdate(opt, queryContext);
    Telemetry.stepUpdated(this);
  }

  get isTrigger(): boolean {
    return this.type === 'trigger';
  }

  get isAction(): boolean {
    return this.type === 'action';
  }

  async getApp() {
    if (!this.appKey) return null;

    return await App.findOneByKey(this.appKey);
  }

  async getLastExecutionStep() {
    const lastExecutionStep = await this.$relatedQuery('executionSteps')
      .orderBy('created_at', 'desc')
      .limit(1)
      .first();

    return lastExecutionStep;
  }

  async getNextStep() {
    const flow = await this.$relatedQuery('flow');

    return await flow
      .$relatedQuery('steps')
      .findOne({ position: this.position + 1 });
  }

  async getTriggerCommand() {
    const { appKey, key, isTrigger } = this;
    if (!isTrigger || !appKey || !key) return null;

    const app = await App.findOneByKey(appKey);
    const command = app.triggers.find((trigger) => trigger.key === key);

    return command;
  }

  async getActionCommand() {
    const { appKey, key, isAction } = this;
    if (!isAction || !appKey || !key) return null;

    const app = await App.findOneByKey(appKey);
    const command = app.actions.find((action) => action.key === key);

    return command;
  }

  async getSetupFields() {
    let setupSupsteps;

    if (this.isTrigger) {
      setupSupsteps = (await this.getTriggerCommand()).substeps;
    } else {
      setupSupsteps = (await this.getActionCommand()).substeps;
    }

    const existingArguments = setupSupsteps.find(
      (substep) => substep.key === 'chooseTrigger'
    ).arguments;

    return existingArguments;
  }

  async updateWebhookUrl() {
    if (this.isAction) return this;

    const payload = {
      webhookPath: await this.computeWebhookPath(),
    };

    await this.$query().patchAndFetch(payload);

    return this;
  }
}

export default Step;
