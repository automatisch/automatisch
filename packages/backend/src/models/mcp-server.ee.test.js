import { describe, it, expect, vi } from 'vitest';
import Crypto from 'node:crypto';

import Base from '@/models/base.js';
import User from '@/models/user.js';
import McpTool from '@/models/mcp-tool.ee.js';
import McpToolExecutions from '@/models/mcp-tool-execution.ee.js';
import McpSession from '@/models/mcp-session.ee.js';
import McpServer from '@/models/mcp-server.ee.js';
import appConfig from '@/config/app.js';
import { createUser } from '@/factories/user.js';
import { createMcpServer } from '@/factories/mcp-server.js';
import { createMcpTool } from '@/factories/mcp-tool.js';
import { createMcpSession } from '@/factories/mcp-session.js';
import { createConnection } from '@/factories/connection.js';
import { createFlow } from '@/factories/flow.js';
import { createStep } from '@/factories/step.js';

describe('McpServer model', () => {
  it('tableName should return correct name', () => {
    expect(McpServer.tableName).toBe('mcp_servers');
  });

  it('jsonSchema should have correct validations', () => {
    expect(McpServer.jsonSchema).toMatchSnapshot();
  });

  it('relationMappings should return correct associations', () => {
    const relationMappings = McpServer.relationMappings();

    const expectedRelations = {
      mcpToolExecutions: {
        join: {
          from: 'mcp_servers.id',
          through: {
            from: 'mcp_tools.mcp_server_id',
            to: 'mcp_tools.id',
          },
          to: 'mcp_tool_executions.mcp_tool_id',
        },
        modelClass: McpToolExecutions,
        relation: Base.ManyToManyRelation,
      },
      mcpSessions: {
        relation: Base.HasManyRelation,
        modelClass: McpSession,
        join: {
          from: 'mcp_servers.id',
          to: 'mcp_sessions.mcp_server_id',
        },
      },
      tools: {
        relation: Base.HasManyRelation,
        modelClass: McpTool,
        join: {
          from: 'mcp_servers.id',
          to: 'mcp_tools.mcp_server_id',
        },
      },
      user: {
        relation: Base.BelongsToOneRelation,
        modelClass: User,
        join: {
          from: 'users.id',
          to: 'mcp_servers.user_id',
        },
      },
    };

    expect(relationMappings).toStrictEqual(expectedRelations);
  });

  describe('generateToken', () => {
    it('should generate a UUID token', () => {
      const mcpServer = new McpServer();
      const testUuid = '550e8400-e29b-41d4-a716-446655440000';

      vi.spyOn(Crypto, 'randomUUID').mockReturnValue(testUuid);
      mcpServer.generateToken();

      expect(mcpServer.token).toBe(testUuid);
    });
  });

  describe('rotateToken', () => {
    it('should generate new token and update in database', async () => {
      const user = await createUser();

      const mcpServer = await createMcpServer({
        userId: user.id,
        name: 'Test Server',
      });

      const newToken = '550e8400-e29b-41d4-a716-446655440001';
      vi.spyOn(Crypto, 'randomUUID').mockReturnValue(newToken);

      const updatedServer = await mcpServer.rotateToken();
      const refetchedServer = await McpServer.query().findById(mcpServer.id);

      expect(updatedServer.token).toBe(newToken);
      expect(refetchedServer.token).toBe(newToken);
    });
  });

  describe('serverUrl', () => {
    it('should return correct server URL with token', () => {
      const mcpServer = new McpServer();
      const testToken = '550e8400-e29b-41d4-a716-446655440003';
      mcpServer.token = testToken;

      const expectedUrl = `${appConfig.baseUrl}/api/v1/mcp/${testToken}`;
      expect(mcpServer.serverUrl).toBe(expectedUrl);
    });
  });

  describe('createTool', () => {
    it('should create new tool when none exists', async () => {
      const notifyToolsListChangedSpy = vi
        .spyOn(McpSession, 'notifyToolsListChanged')
        .mockResolvedValue();

      const user = await createUser();

      const mcpServer = await createMcpServer({
        userId: user.id,
        name: 'Test Server',
      });

      const connection = await createConnection({
        userId: user.id,
      });

      const toolData = {
        connectionId: connection.id,
        appKey: 'slack',
        action: 'sendMessage',
      };

      const mcpTool = await mcpServer.createTool(toolData);

      expect(mcpTool.mcpServerId).toBe(mcpServer.id);
      expect(mcpTool.connectionId).toBe(connection.id);
      expect(mcpTool.appKey).toBe('slack');
      expect(mcpTool.action).toBe('sendMessage');
      expect(mcpTool.type).toBe('app');
      expect(mcpTool.flowId).toBeNull();
      expect(notifyToolsListChangedSpy).toHaveBeenCalledWith(mcpServer.id);
    });

    it('should create flow-type tool without connectionId', async () => {
      const notifyToolsListChangedSpy = vi
        .spyOn(McpSession, 'notifyToolsListChanged')
        .mockResolvedValue();

      const user = await createUser();

      const mcpServer = await createMcpServer({
        userId: user.id,
        name: 'Test Server',
      });

      const flow = await createFlow({
        userId: user.id,
        name: 'Test Flow',
      });

      await createStep({
        type: 'trigger',
        flowId: flow.id,
        appKey: 'mcp',
        key: 'mcpTool',
        parameters: {
          toolName: 'test_flow_tool',
          toolDescription: 'A test flow tool',
        },
      });

      const toolData = {
        type: 'flow',
        flowId: flow.id,
      };

      const mcpTool = await mcpServer.createTool(toolData);

      expect(mcpTool.mcpServerId).toBe(mcpServer.id);
      expect(mcpTool.flowId).toBe(flow.id);
      expect(mcpTool.type).toBe('flow');
      expect(mcpTool.connectionId).toBeNull();
      expect(mcpTool.appKey).toBeNull();
      expect(mcpTool.action).toBe('test_flow_tool');
      expect(notifyToolsListChangedSpy).toHaveBeenCalledWith(mcpServer.id);
    });
  });

  describe('deleteTool', () => {
    it('should delete existing tool and notify sessions', async () => {
      const notifyToolsListChangedSpy = vi
        .spyOn(McpSession, 'notifyToolsListChanged')
        .mockResolvedValue();

      const user = await createUser();

      const mcpServer = await createMcpServer({
        userId: user.id,
        name: 'Test Server',
      });

      const connection = await createConnection({
        userId: user.id,
      });

      const mcpTool = await createMcpTool({
        mcpServerId: mcpServer.id,
        connectionId: connection.id,
        appKey: 'slack',
        action: 'sendMessage',
      });

      const deletedTool = await mcpServer.deleteTool(mcpTool.id);

      expect(deletedTool.id).toBe(mcpTool.id);
      expect(notifyToolsListChangedSpy).toHaveBeenCalledWith(mcpServer.id);

      // Verify tool was deleted from database
      const refetchedTool = await McpTool.query().findById(mcpTool.id);
      expect(refetchedTool).toBeUndefined();
    });

    it('should throw error when tool not found', async () => {
      const user = await createUser();
      const nonExistentId = Crypto.randomUUID();

      const mcpServer = await createMcpServer({
        userId: user.id,
        name: 'Test Server',
      });

      await expect(mcpServer.deleteTool(nonExistentId)).rejects.toThrow();
    });
  });

  describe('delete', () => {
    it('should delete all related tools before deleting server', async () => {
      const user = await createUser();

      const mcpServer = await createMcpServer({
        userId: user.id,
        name: 'Test Server',
      });

      const connection = await createConnection({
        userId: user.id,
      });

      const tool1 = await createMcpTool({
        mcpServerId: mcpServer.id,
        connectionId: connection.id,
        appKey: 'slack',
        action: 'sendMessage',
      });

      const tool2 = await createMcpTool({
        mcpServerId: mcpServer.id,
        connectionId: connection.id,
        appKey: 'github',
        action: 'createIssue',
      });

      const deletedServer = await mcpServer.delete();

      expect(deletedServer.id).toBe(mcpServer.id);

      // Verify tools were deleted
      const deletedTool1 = await McpTool.query().findById(tool1.id);
      const deletedTool2 = await McpTool.query().findById(tool2.id);
      expect(deletedTool1).toBeUndefined();
      expect(deletedTool2).toBeUndefined();

      // Verify server was deleted
      const deletedServer2 = await McpServer.query().findById(mcpServer.id);
      expect(deletedServer2).toBeUndefined();
    });

    it('should delete server even when no tools exist', async () => {
      const user = await createUser();

      const mcpServer = await createMcpServer({
        userId: user.id,
        name: 'Test Server',
      });

      const deletedServer = await mcpServer.delete();

      expect(deletedServer.id).toBe(mcpServer.id);

      // Verify server was deleted
      const refetchedServer = await McpServer.query().findById(mcpServer.id);
      expect(refetchedServer).toBeUndefined();
    });

    it('should call terminateSessions', async () => {
      const terminateSessionsSpy = vi
        .spyOn(McpServer.prototype, 'terminateSessions')
        .mockResolvedValue();

      const user = await createUser();

      const mcpServer = await createMcpServer({
        userId: user.id,
        name: 'Test Server',
      });

      await mcpServer.delete();

      expect(terminateSessionsSpy).toHaveBeenCalled();
    });

    it('should delete all related sessions before deleting server', async () => {
      const user = await createUser();

      const mcpServer = await createMcpServer({
        userId: user.id,
        name: 'Test Server',
      });

      const session1 = await createMcpSession({
        mcpServerId: mcpServer.id,
      });

      const session2 = await createMcpSession({
        mcpServerId: mcpServer.id,
      });

      const session3 = await createMcpSession({
        mcpServerId: mcpServer.id,
      });

      const deletedServer = await mcpServer.delete();

      expect(deletedServer.id).toBe(mcpServer.id);

      const deletedSession1 = await McpSession.query().findById(session1.id);
      const deletedSession2 = await McpSession.query().findById(session2.id);
      const deletedSession3 = await McpSession.query().findById(session3.id);

      expect(deletedSession1).toBeUndefined();
      expect(deletedSession2).toBeUndefined();
      expect(deletedSession3).toBeUndefined();

      const refetchedServer = await McpServer.query().findById(mcpServer.id);
      expect(refetchedServer).toBeUndefined();
    });
  });

  describe('notifyToolsListChanged', () => {
    it('should call mcp session with server id', async () => {
      const notifyToolsListChangedSpy = vi
        .spyOn(McpSession, 'notifyToolsListChanged')
        .mockResolvedValue();

      const user = await createUser();
      const mcpServer = await createMcpServer({
        userId: user.id,
        name: 'Test Server',
      });

      await mcpServer.notifyToolsListChanged();

      expect(notifyToolsListChangedSpy).toHaveBeenCalledWith(mcpServer.id);
    });
  });

  describe('$beforeInsert', () => {
    it('should generate token before inserting', async () => {
      const user = await createUser();
      const testUuid = '550e8400-e29b-41d4-a716-446655440002';

      vi.spyOn(Crypto, 'randomUUID').mockReturnValue(testUuid);

      const mcpServer = await McpServer.query().insertAndFetch({
        userId: user.id,
        name: 'Test Server',
      });

      expect(mcpServer.token).toBe(testUuid);
    });
  });

  describe('createTool', () => {
    it('should throw error when creating duplicate app tool', async () => {
      const user = await createUser();
      const connection = await createConnection({
        userId: user.id,
        key: 'slack',
      });

      const mcpServer = await createMcpServer({
        userId: user.id,
        name: 'Test Server',
      });

      // Create first tool
      await mcpServer.createTool({
        type: 'app',
        appKey: 'slack',
        action: 'sendMessage',
        connectionId: connection.id,
      });

      // Try to create duplicate tool
      await expect(
        mcpServer.createTool({
          type: 'app',
          appKey: 'slack',
          action: 'sendMessage',
          connectionId: connection.id,
        })
      ).rejects.toThrow('Tool with the same app and action already exists');
    });

    it('should throw error when creating duplicate flow tool', async () => {
      const user = await createUser();
      const flow = await createFlow({ userId: user.id });

      // Create trigger step for the flow
      await createStep({
        flowId: flow.id,
        type: 'trigger',
        appKey: 'mcp',
        key: 'mcpTool',
        parameters: {
          toolName: 'test_flow_tool',
          toolDescription: 'Test flow tool',
        },
      });

      const mcpServer = await createMcpServer({
        userId: user.id,
        name: 'Test Server',
      });

      // Create first flow tool
      await mcpServer.createTool({
        type: 'flow',
        flowId: flow.id,
      });

      // Try to create duplicate flow tool
      await expect(
        mcpServer.createTool({
          type: 'flow',
          flowId: flow.id,
        })
      ).rejects.toThrow('Flow tool with the same flow already exists');
    });
  });
});
