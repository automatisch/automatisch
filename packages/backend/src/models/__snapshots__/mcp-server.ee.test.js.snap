// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`McpServer model > jsonSchema should have correct validations 1`] = `
{
  "properties": {
    "createdAt": {
      "type": "string",
    },
    "id": {
      "format": "uuid",
      "type": "string",
    },
    "name": {
      "type": "string",
    },
    "token": {
      "format": "uuid",
      "type": "string",
    },
    "updatedAt": {
      "type": "string",
    },
    "userId": {
      "format": "uuid",
      "type": "string",
    },
  },
  "required": [
    "name",
  ],
  "type": "object",
}
`;
